### PR TITLE
Fix!: Integrate dbt macros into creation of model and source names

### PR DESCRIPTION
**Pull request by crericha**

### CODE REVIEW

File: sqlmesh/dbt/basemodel.py [hunk 1]
```diff
@@ -170,17 +170,29 @@ def config_name(self) -> str:
...
+        """
+        Get the sqlmesh model name
+
+        Returns:
+            The sqlmesh model name
+        """
+        # TODO add back in conditional database
+        # database = self.database if self.database != context.default_database else None
+        database = self.database
+        return ".".join(part for part in (database, self.table_schema, self.table_name) if part)
```
Comment (izeigerman) [inline sqlmesh/dbt/basemodel.py:194]: Why is this equivalent to `full_sql_name`? Should we just call that with TODO?

File: sqlmesh/dbt/loader.py [hunk 2]
```diff
@@ -150,7 +150,10 @@ def _load_project(self) -> Project:
             )
 
         for package_name, macro_infos in context.manifest.all_macros.items():
-            context.jinja_macros.add_macros(macro_infos, package=package_name)
+            context.add_macros(macro_infos, package=package_name)
+
+        # TODO this needs to make its way into engine adapter's default database
+        context.default_database = context.render("{{ generate_database_name() }}")
```
Comment (izeigerman) [inline sqlmesh/dbt/loader.py:156]: I'd rather call `jinja_macros` directly and not bother with `Context.add_macros`
```
context.jinja_macros.build_macro(MacroReference(name="generate_database_name"))()
```
Feel free to have a helper for this in the context. Or better yet just make it a proeprty method of the `Context` instead:
```
@property
def default_database(self):
    return self.jinja_macros.build_macro(MacroReference(name="generate_database_name"))()
```

Comment (izeigerman) [reply sqlmesh/dbt/loader.py:156]: This way you can also do a proper error handling if the `generate_database_name` macro is missing.

File: sqlmesh/dbt/model.py [hunk 3]
```diff
@@ -172,7 +172,7 @@ def model_kind(self, target: TargetConfig) -> ModelKind:
                     logger.warning(
                         "SQLMesh incremental by time strategy is not compatible with '%s' incremental strategy in model '%s'. Supported strategies include %s.",
                         strategy,
-                        self.sql_name,
+                        self.full_sql_name,
```
Comment (izeigerman) [inline sqlmesh/dbt/model.py:175]: Why full name in all error messages? I thought the canonical name was `sql_name`

File: sqlmesh/dbt/basemodel.py [hunk 4]
```diff
@@ -170,17 +170,29 @@ def config_name(self) -> str:
...
 
         Returns:
-            The sqlmesh model name
+            The fully qualified model name
         """
         return ".".join(
             part for part in (self.database, self.table_schema, self.table_name) if part
         )
 
+    def sql_name(self, context: DbtContext) -> str:
```
Comment (izeigerman) [inline sqlmesh/dbt/basemodel.py:184]: Should we call this `canonical_name` for clarity?

File: sqlmesh/dbt/basemodel.py [hunk 5]
```diff
@@ -170,17 +170,29 @@ def config_name(self) -> str:
         return f"{self.package_name}.{self.name}"
 
     @property
-    def sql_name(self) -> str:
+    def full_sql_name(self) -> str:
```
Comment (izeigerman) [inline sqlmesh/dbt/basemodel.py:173]: `fqn`?

File: sqlmesh/dbt/source.py [hunk 6]
```diff
@@ -55,9 +60,13 @@ def table_name(self) -> t.Optional[str]:
     def config_name(self) -> str:
         return f"{self.source_name_}.{self.name}"
 
-    @property
-    def sql_name(self) -> str:
-        return ".".join(part for part in (self.database, self.schema_, self.table_name) if part)
+    def sql_name(self, context: DbtContext) -> str:
+        if self._sql_name is None:
+            self._sql_name = context.render(
```
Comment (izeigerman) [inline sqlmesh/dbt/source.py:65]: Can we use `build_macro` instead?

File: sqlmesh/dbt/basemodel.py (hunk 1 above)
Comment (crericha) [reply sqlmesh/dbt/basemodel.py:194]: Once the renderer supports translating fqn or model name to physical table, the intention is to use the commented out logic. @eakmanrq is working on adding the renderer logic.

File: sqlmesh/dbt/loader.py (hunk 2 above)
Comment (crericha) [reply sqlmesh/dbt/loader.py:156]: Good idea!

File: sqlmesh/dbt/model.py (hunk 3 above)
Comment (crericha) [reply sqlmesh/dbt/model.py:175]: `sql_name` requires the context object. For loading, I thought using the fqn could be a substitute. The other option I considered was package.model (e.g. sushi.items). What do you think?

File: sqlmesh/dbt/basemodel.py (hunk 4 above)
Comment (crericha) [reply sqlmesh/dbt/basemodel.py:184]: I have no opinion. I'm good with whatever. It used to be called `sqlmesh_name` I believe.

Comment (izeigerman) [reply sqlmesh/dbt/basemodel.py:184]: IMHO `sql_name` outlived its relevance

Comment (crericha) [reply sqlmesh/dbt/basemodel.py:184]: Up and onwards!

File: sqlmesh/dbt/adapter.py [hunk 7]
```diff
@@ -295,41 +302,42 @@ def execute(
...
-            return None
-        return identifier
+        identifier = self._map_table_name(self._normalize(self._relation_to_table(relation))).name
+        return identifier if identifier else None
 
-    def _map_table_name(
-        self, database: t.Optional[str], schema: t.Optional[str], identifier: t.Optional[str]
-    ) -> exp.Table:
-        name = ".".join(p for p in (database, schema, identifier) if p is not None)
+    def _map_table_name(self, table: exp.Table) -> exp.Table:
```
Comment (tobymao) [inline sqlmesh/dbt/adapter.py:312]: why isn't normalization happen here?

File: sqlmesh/dbt/adapter.py [hunk 8]
```diff
@@ -295,41 +302,42 @@ def execute(
...
 
     def _relation_to_table(self, relation: BaseRelation) -> exp.Table:
-        assert relation.identifier is not None
-        return exp.table_(relation.identifier, db=relation.schema, catalog=relation.database)
+        table = exp.to_table(relation.render(), dialect=self.engine_adapter.dialect)
+        return exp.to_table(relation.render(), dialect=self.engine_adapter.dialect)
+
+    def _table_to_relation(self, table: exp.Table) -> BaseRelation:
+        return self.relation_type.create(
+            database=table.catalog or None,
```
Comment (tobymao) [inline sqlmesh/dbt/adapter.py:328]: why are you calling table.catalog? you lose quotes if you do that

Comment (crericha) [reply sqlmesh/dbt/adapter.py:328]: mimics what dbt does

File: sqlmesh/dbt/adapter.py (hunk 7 above)
Comment (crericha) [reply sqlmesh/dbt/adapter.py:312]: I kept normalization and mapping separate, so it's clearer what's happening; however, we could in theory have a relation to mapping method.

File: sqlmesh/dbt/basemodel.py [hunk 9]
```diff
@@ -170,17 +174,33 @@ def config_name(self) -> str:
...
+
+        Returns:
+            The sqlmesh model name
+        """
+        if not self._canonical_name:
+            relation = context.create_relation(self.relation_info)
+            # TODO add back in conditional database
+            # if relation.database == context.default_database:
+            #    relation.database = None
+            table = normalize_identifiers(exp.to_table(relation.render(), dialect=context.dialect))
```
Comment (tobymao) [inline sqlmesh/dbt/basemodel.py:200]: use the normalize_model_name method

File: sqlmesh/dbt/source.py [hunk 10]
```diff
@@ -55,9 +63,25 @@ def table_name(self) -> t.Optional[str]:
...
+        if self._canonical_name is None:
+            source = context.get_callable_macro("source")
+            if not source:
+                raise ConfigError("'source' macro not found.")
+            try:
+                relation = source(self.source_name_, self.name)
+                # TODO add back in conditional database
+                # if relation.database == context.default_database:
+                #    relation.database = None
+                table = normalize_identifiers(
```
Comment (tobymao) [inline sqlmesh/dbt/source.py:76]: normalize_model_name

**Comment by crericha** [comment]
The content of this PR has been incorporated into PR #1574.

**Pull request closed without merging**